gtsam/slam/PlanarProjectionFactor.h

```diff
@@ -0,0 +1,318 @@
+
+    private:
+        static const Pose3 CAM_COORD;
+    };
+
+    // camera "zero" is facing +z; this turns it to face +x
+    const Pose3 PlanarProjectionFactorBase::CAM_COORD = Pose3(
+        Rot3(0, 0, 1,//
+            -1, 0, 0, //
+            0, -1, 0),
+        Vector3(0, 0, 0)
+    );
+
+    /**
+     * @class PlanarProjectionFactor1
+     * @brief One variable: the pose.
+     * Landmark, camera offset, camera calibration are constant.
+     * This is intended for localization within a known map.
+     */
+    class PlanarProjectionFactor1
+        : public PlanarProjectionFactorBase, public NoiseModelFactorN<Pose2> {
+    public:
+        typedef NoiseModelFactorN<Pose2> Base;
+        using Base::evaluateError;
+        PlanarProjectionFactor1() {}
+
+        ~PlanarProjectionFactor1() override {}
+
+        /// @return a deep copy of this factor
+        gtsam::NonlinearFactor::shared_ptr clone() const override {
+            return std::static_pointer_cast<gtsam::NonlinearFactor>(
+                gtsam::NonlinearFactor::shared_ptr(new PlanarProjectionFactor1(*this)));
+        }
+
+
+        /**
+         * @param landmark point3 in the world
+         * @param measured corresponding point2 in the camera frame
+         * @param offset constant camera offset from pose
+         * @param calib constant camera calibration
+         * @param model stddev of the measurements, ~one pixel?
+         * @param poseKey index of the robot pose2 in the z=0 plane
+         */
+        PlanarProjectionFactor1(
+            const Point3& landmark,
+            const Point2& measured,
+            const Pose3& offset,
+            const Cal3DS2& calib,
+            const SharedNoiseModel& model,
+            Key poseKey)
+            : PlanarProjectionFactorBase(measured),
+            NoiseModelFactorN(model, poseKey),
+            landmark_(landmark),
+            offset_(offset),
+            calib_(calib) {
+            assert(2 == model->dim());
+        }
+
+        /**
+         * @param pose estimated pose2
+         * @param H1 pose jacobian
+         */
+        Vector evaluateError(
+            const Pose2& pose,
+            OptionalMatrixType H1 = OptionalNone) const override {
+            return h(landmark_, pose, offset_, calib_, {}, H1, {}, {}) - measured_;
+        }
+
+    private:
+        Point3 landmark_; // landmark
+        Pose3 offset_; // camera offset to robot pose
+        Cal3DS2 calib_; // camera calibration
+    };
+
+    template<>
+    struct traits<PlanarProjectionFactor1> :
+        public Testable<PlanarProjectionFactor1> {};
+
+    /**
+     * @class PlanarProjectionFactor2
+     * @brief Two unknowns: the pose and the landmark.
+     * Camera offset and calibration are constant.
+     * This is similar to GeneralSFMFactor, used for SLAM.
+    */
+    class PlanarProjectionFactor2 : public PlanarProjectionFactorBase, public NoiseModelFactorN<Point3, Pose2> {
```

Comment:
I'd reverse and do Pose Point. I think that's what we do everywhere?

Reply:
done